quinn-proto/src/connection/mod.rs

```diff
@@ -760,10 +784,104 @@ where
+                    }
+
+                    let next_checkpoint = self.cids_timeline.pop_front();
+                    let next_seq = next_checkpoint.unwrap().sequence + 1;
+                    let current_seq = self.retire_cid_seq;
+                    //  Endpoints SHOULD NOT issue updates of the Retire Prior To field
+                    //  before receiving RETIRE_CONNECTION_ID frames that retire all
+                    //  connection IDs indicated by the previous Retire Prior To value.
+                    // https://tools.ietf.org/html/draft-ietf-quic-transport-29#section-5.1.2
+                    if next_seq > self.retire_cid_seq && !ids_found {
+                        self.retire_cid_seq = next_seq;
+                    }
+                    // An endpoint MUST NOT
+                    // provide more connection IDs than the peer's limit.
+                    //
+                    // An endpoint MAY
+                    // send connection IDs that temporarily exceed a peer's limit if the
+                    // NEW_CONNECTION_ID frame also requires the retirement of any excess,
+                    // by including a sufficiently large value in the Retire Prior To field.
+                    // https://tools.ietf.org/html/draft-ietf-quic-transport-29#section-5.1.1
+                    // In this case, `retire prior to` must increase to accommodate new cid
+                    let must_increase_retire_cid =
+                        self.cids_active_seq.len() as u64 >= self.peer_params.issue_cids_limit();
+                    if must_increase_retire_cid
+                        && self.retire_cid_seq - current_seq < num_pushed_cids
+                    {
+                        self.retire_cid_seq = current_seq + num_pushed_cids;
+                    }
+
+                    // In case we already received RETIRE_CONNECTION_ID before
+                    let mut push_new_cid = false;
+                    for id in current_seq..self.retire_cid_seq {
+                        push_new_cid = push_new_cid || self.cids_active_seq.contains(&id);
+                    }
+
+                    if push_new_cid {
+                        if !self.state.is_closed() {
+                            trace!(
+                                "push a new cid to peer RETIRE_PRIOR_TO field {}",
+                                self.retire_cid_seq
+                            );
+                            self.endpoint_events
+                                .push_back(EndpointEventInner::NeedIdentifiers(
+                                    now,
+                                    num_pushed_cids,
+                                ));
+                        }
+                    } else {
+                        // We don't need to push new cid as peer has actively retired them already
+                        self.reset_cid_timer(None);
+                    }
+                }
```

Comment:
Overall I find the logic here difficult to follow. Can you add comments expanding on intentions throughout? Citing the spec is useful for explaining why we need to do something, but the part I'm having trouble with is understanding what we're actually doing here.

Reply:
Ack.